storage/spanstore/downsampling_writer.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanstore
+
+import (
+	"hash"
+	"hash/fnv"
+	"math"
+	"sync"
+
+	"github.com/uber/jaeger-lib/metrics"
+
+	"github.com/jaegertracing/jaeger/model"
+)
+
+const (
+	traceIDByteSize = 16
+)
+
+// DownSamplingWriter is a span Writer that drops spans with a predefined downSamplingRatio.
+type DownSamplingWriter struct {
+	spanWriter   Writer
+	threshold    uint64
+	hashSalt     []byte
+	poolInstance *poolSingleton
+	metrics      *downSamplingWriterMetrics
+}
+
+// downSamplingWriterMetrics is metrics that keeping track of total number of dropped spans
+type downSamplingWriterMetrics struct {
+	SpansDropped metrics.Counter `metric:"spans_dropped"`
+}
+
+// Wrap the sync.Pool object within a singleton struct.
+type poolSingleton struct {
+	hashpool *sync.Pool
+	bytepool *sync.Pool
+}
+
+var instance *poolSingleton
+var once sync.Once
+
+// Initialize the singleton if it doesn't exist already.
+func getHashPoolInstance(length int) *poolSingleton {
```

Comment:
It is customary to have constructor follow the declaration. Helper functions go later.